contracts/deposit-contracts/contract/src/detail.rs

```diff
@@ -0,0 +1,110 @@
+/*
+    The utilities found in this file were scraped from other Casper contracts,
+    mainly cep-78 and cep-18.
+    This file is not necessarily due for review, unless breaking changes are suspected.
+*/
+
+use crate::error::DepositError;
+use alloc::vec::Vec;
+use casper_contract::{
+    contract_api::{self, runtime},
+    ext_ffi,
+    unwrap_or_revert::UnwrapOrRevert,
+};
+use casper_types::{
+    api_error, bytesrepr, bytesrepr::FromBytes, system::CallStackElement, ApiError, Key, URef,
+};
+use core::convert::TryInto;
+
+fn call_stack_element_to_address(call_stack_element: CallStackElement) -> Key {
+    match call_stack_element {
+        CallStackElement::Session { account_hash } => Key::from(account_hash),
+        CallStackElement::StoredSession { account_hash, .. } => {
+            // Stored session code acts in account's context, so if stored session wants to interact
+            // with an CEP-18 token caller's address will be used.
+            Key::from(account_hash)
+        }
+        CallStackElement::StoredContract {
+            contract_package_hash,
+            ..
+        } => Key::from(contract_package_hash),
+    }
+}
+
+/// Traverse the callstack
```

Comment:
You are not supposed to use this function at all.

Reply:
Here is part from Casper [documentation](https://docs.casper.network/concepts/callstack/):

> Developers can access the call stack with the [contract_api::runtime::get_call_stack](https://docs.rs/casper-contract/3.0.0/casper_contract/contract_api/runtime/fn.get_call_stack.html) function.
> 
> If session code calls a contract, which in turn calls another contract, then the session code would represent the zeroth entity in the stack, the contract called by the initiating session code would be the first and the contract called by the first contract would be the second.

I think this function is supposed to get **immediate caller**, so effectively _parent_ of current execution context (because we have session code that calls deposit contract). Having explanation in the code would be nice though.

Reply:
I have tested the get immediate caller function many times and it works well in contract context. I only use it in contract entry points so it will retrieve the n-1th entity calling the entry point.